### ISSUE TITLE
Undesired constant folding in Fortran printer can cause numerical issues

### ISSUE THREAD

**Issue by brocksam** [sympy/sympy#25435]
Consider the following real example taken from biomechanics, where `fl_T` is the characteristic equation describing the force-length properties of tendon as a function of normalised tendon length `l_T_tilde`:

```python
from sympy import Add, Mul, Symbol, exp
l_T_tilde = Symbol('l_T_tilde')
fl_T1 = 0.2 * exp(33.93669377311689 * (l_T_tilde - 0.995)) - 0.25
# This evaluates to: fl_T1 = 4.32716793402738e-16*exp(33.9366937731169*l_T_tilde) - 0.25
fl_T2 = Add(Mul(0.2, exp(Mul(33.93669377311689, (l_T_tilde - 0.995), evaluate=False)), evaluate=False),
            -0.25, evaluate=False)
# This evaluates to: fl_T2 = 0.2 * exp(33.93669377311689 * (l_T_tilde - 0.995)) - 0.25
```

Normal operating range for `l_T_tilde` is 0.95 < `l_T_tilde` < 1.05. However, there are situations where where a user might need to evaluate the function in the range 0.0 < `l_T_tilde` < 4.0. Higher values would cause overflow for 32 bit single-precision floats. The formulation of `fl_T1` is less well conditioned numerically for values close to the normal operating range, so the formulation of `fl_T2` is preferable for codegen purposes.

Most code printers print `fl_T2` as is. For example, the `C89CodePrinter` will print the desired C code equivalent to `fl_T2`:

```c
0.20000000000000001*exp(33.936693773116893*(l_T_tilde - 0.995)) - 0.25
```

However, the `FCodePrinter` will evaluate the constants inside the `exp` and print undesirable Fortran code more similar to fl_T1: 

```fortran
      0.2d0*2.1635839670136868d-15*exp(33.936693773116893d0*l_T_tilde) -
     @ 0.25d0
```

This is happening with `FCodePrinter._print_Function`. Note that this method is not overloaded for most other printers which is causing the difference in behaviour. https://github.com/sympy/sympy/blob/db5ca227cb29ca6dcb8da8da6230915f21e927e5/sympy/printing/fortran.py#L303-L311

One option would be amend the implementation of `FCodePrinter._print_Function` to do something sensible when the evaluation of constants results in very large or very small values. Alternative we could leave as is.

**Comment by bjodah** [comment]
I think we should probably make this aggressive rewriting in the printer opt-in or simply remove it unless we suspect it causes a serious backwards compatibility issue? I suppose the question is if we can find a plausible counter-example where the rewriting is advantageous. But even then, we don't want the Fortran printer to deviate much from, as you point out, the C and C++ printers. From that standpoint I think the current behavior is borderline a bug. And as you have found out, working around this involves explicitly setting `evaluate=False`.

In general I think we should try to move magic rewriting out of the printers and instruct users to preprocess their expressions prior to printing them. This kind of constant folding sounds like it might fit under `sympy.codegen.rewriting`?

**Comment by brocksam** [comment]
I agree, these rewrites should either be opt-in or alternatively there could be a base set of rewrites on by default if it is easy for a user to disable/replace these.

I can't see how removing it could cause a backwards compatibility issue as the code generated would still be "correct". Also, with optimisations turned on a compiler would likely do most of these constant folding simplifications itself anyway.

I will explore both removing this behaviour from the Fortran printer so that it's closer to mirroring the C and C++ printers. I really like your suggestion to use `sympy.codegen.rewriting` for these rewrites, I'll explore this too.
